File: solutions/kotlin/sieve/1/src/main/kotlin/Sieve.kt
object Sieve {

    fun primesUpTo(upperBound: Int): List<Int> {
        var numbers = Array<Int>(upperBound){0}
        var n=2
        while(n<=upperBound) {
            for (k in n*n..upperBound step n){
                numbers[k-1] = 1
            }
            n++
        }
        return numbers
            .mapIndexed{idx,v ->
                if(idx>0 && v==0) idx+1 else 0}
            .filter{it>0}.toList()
    }
}

File: solutions/dart/sieve/1/lib/sieve.dart
import 'package:collection/collection.dart';

class Sieve {
  
  final int limit;
  
  Sieve(this.limit);

  List<int> get primes {
    if(limit<2)
      return <int>[];
    var marks =  List<bool>
      .generate(limit, (idx) => idx%2==0,
                growable : false); 
    // idx-1 is the corresponding number
    marks[1] = true; // 2
    int n = 3;
    while(n<=limit){
      if(marks[n-1]){
      for(int k=n*n; k<=limit; k+=n){
        marks[k-1] = false;
      }}
      n++;      
    }    
    //transform marks to numbers,
    // remove non-primes and 1
    return marks 
      .mapIndexed( (idx, mark) 
                  => mark? idx+1 : 0) 
      .where((x) =>x>1).toList();
  }
}
